src/dicomweb_client/uri.py

```diff
@@ -0,0 +1,537 @@
+            sop_instance_uid
+            if sop_instance_uid is not None else self.sop_instance_uid,
+            frames if frames is not None else self.frames,
+            suffix if suffix is not None else self.suffix,
+        )
+
+    @property
+    def parent(self) -> 'URI':
+        """Returns a URI to the "parent" resource.
+
+        Depending on the `type` of the current `URI`, the `URI` of the parent
+        resource is defined as:
+
+        +----------+----------+----------+
+        | Current  | Suffixed | Parent   |
+        +==========+==========+==========+
+        | Service  | N/A      | Service  |
+        +----------+----------+----------+
+        | Study    | No       | Service  |
+        +----------+----------+----------+
+        | Study    | Yes      | Study    |
+        +----------+----------+----------+
+        | Series   | No       | Study    |
+        +----------+----------+----------+
+        | Series   | Yes      | Series   |
+        +----------+----------+----------+
+        | Instance | No       | Series   |
+        +----------+----------+----------+
+        | Instance | Yes      | Instance |
+        +----------+----------+----------+
+        | Frame    | No       | Instance |
+        +----------+----------+----------+
+        | Frame    | Yes      | Frame    |
+        +----------+----------+----------+
+
+        Returns
+        -------
+        URI
+          An instance pointing to the parent resource.
+        """
+        if self.type == URIType.SERVICE or self.suffix is not None:
+            return URI(
+                self.base_url,
+                self.study_instance_uid,
+                self.series_instance_uid,
+                self.sop_instance_uid,
+                self.frames,
+                suffix=None)
+        elif self.type == URIType.STUDY:
+            return URI(self.base_url)
+        elif self.type == URIType.SERIES:
+            return self.study_uri()
+        elif self.type == URIType.INSTANCE:
+            return self.series_uri()
+        else:
+            return self.instance_uri()
+
+    @property
+    def parts(self) -> Tuple[str, ...]:
+        """Returns the sequence of URI components in a `tuple`.
+
+        For example, if the URI is:
+        ``http://srv.com/studies/1.2.3/series/4.5.6/instances/7.8/frames/3,4,5/rendered``
+
+        then the method returns
+        ``('https://srv.com', '1.2.3', '4.5.6', '7.8', '3', '4', '5', 'rendered')``
+
+        Returns
+        -------
+        Tuple[str, ...]
+            Sequence of URI components.
+        """  # noqa
+        frames = (
+            () if self.frames is None else
+            tuple(str(frame_number) for frame_number in self.frames))
+        suffix = None if self.suffix is None else self.suffix.value
+        return tuple(part for part in (
+            self.base_url, self.study_instance_uid, self.series_instance_uid,
+            self.sop_instance_uid, *frames, suffix) if part is not None)
```

Comment:
Playing with this again, I noticed that each frame is returned as an individual part. This was unexpected, since I would consider the sequence of frame numbers one part. What do you think about returning them as a tuple instead (i.e., `tuple(frames)` rather than `*frames`)?

Reply:
Thank you for the question!

This was a bit of a cop-out from my end due to the typing -- it's challenging to have meaningful type-consistency if a single method were to return a sequence of strings, tuples, and more strings (due to the suffix). I felt that it might play into the argument of duck-typing being a "weakness" of Python.

The length of the sequence returned by `parts` was what I initially foresaw the caller would use to interpret its content. Since the introduction of the `suffix` and `frame`, this has no longer been true, e.g., `parts` for a Series URI with Thumbnail will have the same length as for an Instance URI without a suffix. The issue persists with the proposed change as well.

It would be helpful to know the kinds of use cases you had in mind for `parts` in its current or proposed forms.

We could consider the following options (please feel free to amend the list):
1. Return `frames` as a `tuple` in `parts`.
2. Return `frames` as `*frames` in `parts`.
3. Drop `parts` altogether.
4. Change the API of `parts` to a key-value map, that could be directly used to initialize an equivalent instance via `kwargs` when passed to `URI.__init__`.

WDYT?

Reply:
Thanks for the thorough assessment. I would prefer option 3 (drop `parts` altogether). The individual parts are already accessible by name (read only!) on instances of the class itself via the implemented properties.  The `URI` class is thus conceptually and functionally very similar to a `NamedTuple`. Instances of the `URI` are immutable. They can be "updated" via `update()`, but this method actually creates a new instance rather than updating the object itself. This is another major and important difference to `dataclasses`, which create mutable objects.

I think that removing `parts` will also make it easier to refactor the class in the future, e.g., adding new parts such as query parameters, fragments, authority, etc. I am not sure we should be doing this (as you pointed out that this use case is already covered by `urllib.parse.urlparse()`), but keeping that option may not be a bad idea.

Reply:
>  I would prefer option 3 (drop parts altogether).

Done.

> This is another major and important difference to dataclasses, which create mutable objects.

Good point. Quick note that setting [`frozen=True`](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass) in the decorator could achieve this, but this may not be pertinent at the moment. I've included a note in the `URI` class docstring to this effect (i.e., instances are immutable).

> I think that removing parts will also make it easier to refactor the class in the future, e.g., adding new parts such as query parameters, fragments, authority, etc. I am not sure we should be doing this (as you pointed out that this use case is already covered by urllib.parse.urlparse()), but keeping that option may not be a bad idea.

Agreed. I would support and be able to help with including support for query parameters in the next iteration of this class :-)